### ISSUE TITLE
onnx.export make size operations return Tensor instead of int

### ISSUE THREAD

**Issue by oliverdain** [pytorch/pytorch#86481]
### 🐛 Describe the bug

`Tensor.size` [is documented](https://pytorch.org/docs/stable/generated/torch.Tensor.size.html#torch.Tensor.size) to return an `int` if you pass the `dim` argument. Similarly, the Tensor indexing operations return an `int` most of the time. However, when tracing via `torch.onnx.export` these somehow do not work as advertised and instead return a `Tensor` that is a scalar.

Given:

```python
import torch

import torch.nn as nn


class ExampleIdentity(nn.Module):
    def __init__(self) -> None:
        super().__init__()

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        print('Indexing:', type(x.shape[-1]))
        print('Using size:', type(x.size(-1)))
        return x


x = torch.rand((1, 3, 224, 244))
ei = ExampleIdentity()

print('\n\nRunning normally:')
ei(x)

print('\n\nVia onnx.export:')
with open('/tmp/output.onnx', 'wb') as outf:
    torch.onnx.export(ei, x, outf)
```

And running that:

```bash
$ python3.10 /tmp/example.py


Running normally:
Indexing: <class 'int'>
Using size: <class 'int'>


Via onnx.export:
Indexing: <class 'torch.Tensor'>
Using size: <class 'torch.Tensor'>
```

### Versions

```
Collecting environment information...
PyTorch version: 1.12.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.18.4
Libc version: glibc-2.31

Python version: 3.10.6 (main, Sep 20 2022, 13:16:58) [GCC 10.2.1 20210110] (64-bit runtime)
Python platform: Linux-5.10.0-18-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Quadro P520
Nvidia driver version: 470.141.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.12.1
[pip3] torchvision==0.13.1
[conda] Could not collect
```

**Comment by justinchuby** [comment]
Hi! size returns a tensor when the model is traced. This is needed so the traced model can handle dynamic sizes.

**Comment by oliverdain** [comment]
That makes sense but contradicts the documentation and is a bit surprising. I'd vote to make `size` always return a tensor so it's consistent.

**Comment by justinchuby** [comment]
Thanks for your suggestion. Many parts of PyTorch assume this behavior so it may be hard to change. @ezyang do you have insights you would like to share?

**Comment by ezyang** [comment]
We can't make this change. What might work better is to use our new tracing infra where we have a distinct SymInt type for tracing these cases. No confusion with Tensor.

**Comment by ezyang** [comment]
@justinchuby the last time we pitched this to you guys, the blocker was ONNX export is using TorchScript control flow. If we can excise this dependency somehow, a new exporter based on make_fx should be quite feasible.

**Comment by justinchuby** [comment]
Thanks! @BowenBao for awareness
